v2/internal/reconcilers/arm/arm_client_cache.go

```diff
@@ -82,65 +85,130 @@ func (c *armClientCache) lookup(key string) (*armClient, bool) {
+	if err != nil {
+		return nil, "", err
+	} else if client != nil {
+		return client.GenericClient(), client.CredentialFrom(), nil
+	}
 	// Namespaced secret
-	secret, err := c.getSecret(ctx, obj.GetNamespace(), namespacedSecretName)
+	client, err = c.getNamespacedCredential(ctx, obj)
 	if err != nil {
 		return nil, "", err
+	} else if client != nil {
+		return client.GenericClient(), client.CredentialFrom(), nil
 	}
 
-	if secret != nil {
-		nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
-		client, ok := c.lookup(nsName.String())
-		// if client does not already exist, or secret data is changed, we need to get new token.
-		if !ok || !matchSecretData(client.secretData, secret.Data) {
-			credential, subscriptionID, err := c.newCredentialFromSecret(secret, nsName)
-			if err != nil {
-				return nil, "", err
-			}
-
-			newClient, err := genericarmclient.NewGenericClientFromHTTPClient(c.cloudConfig, credential, c.httpClient, subscriptionID, metrics.NewARMClientMetrics())
-			if err != nil {
-				return nil, "", err
-			}
-
-			armClient := newARMClient(newClient, secret.Data, nsName)
-			c.register(armClient)
-			return armClient.GenericClient(), armClient.CredentialFrom(), nil
-		} else {
-			return client.GenericClient(), client.CredentialFrom(), nil
+	// If not found, default is the global client
+	return c.globalClient.GenericClient(), c.globalClient.CredentialFrom(), nil
+}
+
+func (c *armClientCache) getPerResourceCredential(ctx context.Context, obj genruntime.ARMMetaObject) (*armClient, error) {
+	return c.getCredentialFromAnnotation(ctx, obj, perResourceSecretAnnotation)
+}
+
+func (c *armClientCache) getNamespacedCredential(ctx context.Context, obj genruntime.ARMMetaObject) (*armClient, error) {
+	secret, err := c.getSecret(ctx, obj.GetNamespace(), namespacedSecretName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil // Not finding this secret is allowed, allow caller to proceed to higher scope secret
 		}
+		return nil, err
 	}
 
-	// If not found, default would be global
-	return c.globalClient.GenericClient(), c.globalClient.CredentialFrom(), nil
+	armClient, err := c.getARMClientFromSecret(secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return armClient, nil
+}
+
+func (c *armClientCache) getCredentialFromAnnotation(ctx context.Context, obj genruntime.ARMMetaObject, annotation string) (*armClient, error) {
+	credentialFrom, ok := obj.GetAnnotations()[annotation]
+	if !ok {
+		return nil, nil
+	}
+
+	// annotation exists, use specified secret
+	secretNamespacedName := getSecretNameFromAnnotation(credentialFrom, obj.GetNamespace())
+
+	secret, err := c.getSecret(ctx, secretNamespacedName.Namespace, secretNamespacedName.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, core.NewSecretNotFoundError(secretNamespacedName, errors.Errorf("credential Secret %q not found for '%s/%s'", secretNamespacedName.String(), obj.GetNamespace(), obj.GetName()))
```

Comment:
Don't need to do the `errors.Errorf("credential Secret %q not found for '%s/%s'", secretNamespacedName.String(), obj.GetNamespace(), obj.GetName())` anymore, as I think that's all already included in the err, so just do:

`core.NewSecretNotFoundError(secretNamespacedName, err)`. Or if you have to add some extra details, wrap the error so its context isn't lost w/ `errors.Wrapf`.

Reply:
Thanks for mentioning. Have improved the other errors as well.